x/bank/keeper.go

```diff
@@ -8,63 +8,59 @@ import (
 
 const moduleName = "bank"
 
-// CoinKeeper manages transfers between accounts
-type CoinKeeper struct {
-	am sdk.AccountMapper
+func getCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address) sdk.Coins {
+	acc := am.GetAccount(ctx, addr)
+	if acc == nil {
+		return sdk.Coins{}
+	}
+	return acc.GetCoins()
 }
 
-// NewCoinKeeper returns a new CoinKeeper
-func NewCoinKeeper(am sdk.AccountMapper) CoinKeeper {
-	return CoinKeeper{am: am}
+func setCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) sdk.Error {
+	acc := am.GetAccount(ctx, addr)
+	if acc == nil {
+		acc = am.NewAccountWithAddress(ctx, addr)
+	}
+	acc.SetCoins(amt)
+	am.SetAccount(ctx, acc)
+	return nil
 }
 
-// GetCoins returns the coins at the addr.
-func (ck CoinKeeper) GetCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) sdk.Coins {
-	acc := ck.am.GetAccount(ctx, addr)
-	return acc.GetCoins()
+// HasCoins returns whether or not an account has at least amt coins.
+func hasCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) bool {
+	return getCoins(ctx, am, addr).IsGTE(amt)
 }
 
 // SubtractCoins subtracts amt from the coins at the addr.
-func (ck CoinKeeper) SubtractCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
-	acc := ck.am.GetAccount(ctx, addr)
-	if acc == nil {
-		return amt, sdk.ErrUnknownAddress(addr.String())
-	}
-
-	coins := acc.GetCoins()
-	newCoins := coins.Minus(amt)
+func subtractCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	oldCoins := getCoins(ctx, am, addr)
+	newCoins := oldCoins.Minus(amt)
 	if !newCoins.IsNotNegative() {
-		return amt, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", coins, amt))
+		return amt, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
 	}
-
-	acc.SetCoins(newCoins)
-	ck.am.SetAccount(ctx, acc)
-	return newCoins, nil
+	err := setCoins(ctx, am, addr, newCoins)
+	return newCoins, err
 }
 
 // AddCoins adds amt to the coins at the addr.
-func (ck CoinKeeper) AddCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
-	acc := ck.am.GetAccount(ctx, addr)
-	if acc == nil {
-		acc = ck.am.NewAccountWithAddress(ctx, addr)
+func addCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	oldCoins := getCoins(ctx, am, addr)
+	newCoins := oldCoins.Plus(amt)
+	if !newCoins.IsNotNegative() {
```

Comment:
When does this happen - overflow? Should we just panic on overflow? This error would be confusing.

Reply:
Well I was thinking more if someone passes in a negative amount of coins to be added.

Reply:
Oh OK, that's fine I guess.

In general, I find representing negative coins a bit confusing (as you never want anyone to *end up* with negative coins, it's just an intermediate representation, but it uses the same datatype), but maybe that's just me.

Reply:
Agreed.  Negative coins is weird.